### ISSUE TITLE
How to define a Primary or Shard key for the schema_migrations table ?

### ISSUE THREAD

**Issue by Micka33** [rails/rails#21070]
I tried to post [my issue on stackoverflow](http://stackoverflow.com/questions/31701980/ruby-on-rails-with-memsql-table-schema-migrations-must-either-have-a-primary), but it doesn't seems to attract much attention. So I'm trying here, hoping to get answers from experts :D .

I want exactly what the title of this issue says:
 **`to define a Primary or Shard key for the schema_migrations table`**

The use case is the following, I want to migrate my backend from MySQL to MemSQL. It should be pretty straight forward, however i'm encountering the following error.

``` shell
> rake db:migrate
rake aborted!
ActiveRecord::StatementInvalid: Mysql2::Error: Distributed tables must either have a PRIMARY or SHARD key. Visit http://docs.memsql.com/4.0/concepts/porting-apps/ for more information: CREATE TABLE `schema_migrations` (`version` varchar(255) NOT NULL) ENGINE=InnoDB

Mysql2::Error: Distributed tables must either have a PRIMARY or SHARD key. Visit http://docs.memsql.com/4.0/concepts/porting-apps/ for more information
```

The reason of this error is that MemSQL NEED a primary or shard key for each table it handles.

This error is easily reproducible with the following steps
- create a new rails project
- add mysql2 gem
- configure database.yml to memsql ip
- add devise gem
-  add user table with devise
- rake db:create
- rake db:migrate (result of this command below)

**Comment by pixeltrix** [comment]
Does adding one using the `version` column make it work?

**Comment by Micka33** [comment]
The orignal SQL of the schema_migrations table is:

```SQL
DROP TABLE IF EXISTS `schema_migrations`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `schema_migrations` (
  `version` varchar(255) COLLATE utf8_unicode_ci NOT NULL,
  UNIQUE KEY `unique_schema_migrations` (`version`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;
```
it gives the following error:
```SQL
ERROR 1751 (HY000): Distributed tables must either have a PRIMARY or SHARD key. Visit http://docs.memsql.com/4.0/concepts/porting-apps/ for more information
```
When add a primary key the following way, it works fine.
```SQL
memsql> CREATE TABLE `schema_migrations` (
    ->   `version` varchar(255) COLLATE utf8_unicode_ci NOT NULL,
    ->   PRIMARY KEY (`version`),
    ->   UNIQUE KEY `unique_schema_migrations` (`version`)
    -> ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
Query OK, 0 rows affected, 1 warning (0.01 sec)
```

Tell me, if you need more informations.

**Comment by Micka33** [comment]
I haven't tried it again, but this issue will be resolved when there will be a PRIMARY or a SHARD key in the `schema_migrations` table.

**Comment by kamipo** [comment]
Fixed in #23359.
